---
title: "Prepare Environment for Transmission Analyses"
author: "Sebastian Schmidt"
date: "2021-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

options(stringsAsFactors = FALSE)
```

## Prepare Environment

```{r Load Libraries, echo=F, warning=FALSE}
library("tidyverse", warn.conflicts=F, quietly=T)
library("ape", warn.conflicts = F, quietly = T)
```

Set file names and basic parameters.

```{r Set Parameters}
#Preallocate global data structures
PARAM <- list()

#Set relevant folder names
PARAM$folder.Rmd <- paste0(getwd(), "/")
PARAM$folder.base <- gsub("Rmd/", "", PARAM$folder.Rmd)
PARAM$folder.data <- paste0(PARAM$folder.base, "data/")
PARAM$folder.results <- paste0(PARAM$folder.base, "results/")

##########################
#INPUT
##########################
#Sample data
PARAM$file.sample <- paste0(PARAM$folder.data, "data.sample.Rdata")
#Species phenotype data
PARAM$file.phenotype <- paste0(PARAM$folder.data, "data.phenotype.Rdata")
#Species GTDB taxonomy
PARAM$file.taxonomy.gtdb <- paste0(PARAM$folder.data, "data.taxonomy.gtdb.RData")
#Species phylogeny
PARAM$file.tree <- paste0(PARAM$folder.data, "data.tree.Rdata")
#Count table data
PARAM$file.mOTU <- paste0(PARAM$folder.data, "data.mOTU.Rdata")
#Count table data for non-mOTU species clusters
PARAM$file.mOTU.ANI <- paste0(PARAM$folder.data, "data.coverage.COG.MAG.Rdata")
#Count table and beta div data
PARAM$file.beta_div <- paste0(PARAM$folder.data, "data.beta_div.Rdata")
#Pre-processed allele distances
PARAM$file.allele_distances <- paste0(PARAM$folder.data, "data.allele_distances.FMT.Rdata")
#Pre-processed allele diversities
PARAM$file.allele_diversity <- paste0(PARAM$folder.data, "data.allele_diversity.Rdata")
#GMGC-based KO profiles
PARAM$file.gmgc_ko <- paste0(PARAM$folder.data, "data.gmgc.KO.Rdata")
#eggNOG & KEGG data
PARAM$file.eggnog_kegg <- paste0(PARAM$folder.data, "data.eggnog_kegg.Rdata")
PARAM$file.eggnog45_kegg <- paste0(PARAM$folder.data, "data.eggnog45_kegg.Rdata")
#KEGG KO complementarity data
PARAM$file.ko_complementarity <- paste0(PARAM$folder.data, "data.ko_complementarity.Rdata")
#Pre-processed GMM profiles per sample (via GMGC)
PARAM$file.gmm_gmgc <- paste0(PARAM$folder.data, "data.gmm.gmgc.Rdata")
PARAM$file.gmm_gmgc_wide <- paste0(PARAM$folder.data, "data.gmm.gmgc.wide.Rdata")
#Pre-processed GMM profiles and phenotype data per species
PARAM$file.gmm_species <- paste0(PARAM$folder.data, "data.gmm.by_species.Rdata")
#Pre-processed data on transmission quantification
PARAM$file.data_transmission <- paste0(PARAM$folder.data, "data.transmission_snv.Rdata")
#PARAM$file.data_transmission_bg <- paste0(PARAM$folder.data, "data.transmission_snv_bg.Rdata")
#Pre-quantified transmission results
PARAM$file.data_transmission.scored <- paste0(PARAM$folder.data, "data.transmission_snv.scored.Rdata")
#PARAM$file.data_transmission_bg.scored <- paste0(PARAM$folder.data, "data.transmission_snv_bg.scored.Rdata")
##########################


##########################
#OUTPUT
##########################
#Consolidated transmission results, focus on engraftment
PARAM$file.data_transmission.scored.consolidated <- paste0(PARAM$folder.data, "data.transmission_snv.scored.consolidated.Rdata")
##########################
```

Load data.

```{r}
#load(PARAM$file.sample)
load(PARAM$file.sample)
load(PARAM$file.phenotype)
load(PARAM$file.taxonomy.gtdb)
load(PARAM$file.data_transmission)
#load(PARAM$file.data_transmission_bg)
load(PARAM$file.data_transmission.scored)
#load(PARAM$file.data_transmission_bg.scored)
load(PARAM$file.mOTU)
load(PARAM$file.mOTU.ANI)
load(PARAM$file.tree)
load(PARAM$file.beta_div)
load(PARAM$file.allele_distances)
load(PARAM$file.allele_diversity)
load(PARAM$file.gmm_gmgc)
load(PARAM$file.gmm_species)
load(PARAM$file.ko_complementarity)
```

Define ANI clusters to drop because they are redundant with ref_mOTU_1383 (Collinsella aerofaciens).

```{r}
drop.Collinsella <- c("ANI_AL_95_00203", "ANI_AL_95_00204", "ANI_AL_95_00205", "ANI_AL_95_00206", "ANI_AL_95_00207", "ANI_AL_95_00208", "ANI_AL_95_00209", "ANI_AL_95_00210", "ANI_AL_95_00211", "ANI_AL_95_00212", "ANI_AL_95_00213", "ANI_AL_95_00214", "ANI_AL_95_00215", "ANI_AL_95_00216", "ANI_AL_95_00217", "ANI_AL_95_00218", "ANI_AL_95_00219", "ANI_AL_95_00220", "ANI_AL_95_00221", "ANI_AL_95_00222", "ANI_AL_95_00223", "ANI_AL_95_00224", "ANI_AL_95_00225", "ANI_AL_95_00226", "ANI_AL_95_00227", "ANI_AL_95_00228", "ANI_AL_95_00229", "ANI_AL_95_00230", "ANI_AL_95_00231", "ANI_AL_95_00232", "ANI_AL_95_00233", "ANI_AL_95_00234", "ANI_AL_95_00235", "ANI_AL_95_00236", "ANI_AL_95_00237", "ANI_AL_95_00238", "ANI_AL_95_00239", "ANI_AL_95_00240", "ANI_AL_95_00241", "ANI_AL_95_00242", "ANI_AL_95_00243")
```

Set plotting parameters.

```{r}
PLOT <- list()

#Set base font families
PLOT$font <- list()
PLOT$font$base_family = "Helvetica"
PLOT$font$size.base = 8
PLOT$font$size.title = font.base_size = 10
PLOT$font$size.text = 8

PLOT$point.size = 2

#Define colour scopes
PLOT$colour <- list()

#FMT outcome colours
PLOT$colour$outcome <- list()
PLOT$colour$outcome$species_lost <- "#fbb4ae" #"#ffff99"
PLOT$colour$outcome$species_retained <- "#b15928"
PLOT$colour$outcome$rejection_consp <- "#ffff33"
PLOT$colour$outcome$influx_consp <- "#c51b7d"
PLOT$colour$outcome$coexistence <- "#fdbf6f"
PLOT$colour$outcome$engraftment_consp <- "#1f78b4"
PLOT$colour$outcome$engraftment_novel <- "#a6cee3"
PLOT$colour$outcome$influx_novel <- "#de77ae"
PLOT$colour$outcome$rejection_novel <- "#deebf7"

#Association term colours
PLOT$colour$term <- list()
PLOT$colour$term$min_depth <- "#f0f0f0"
PLOT$colour$term$indication.rcdi <- "#1f78b4"
PLOT$colour$term$indication.ibd <- "#a6cee3"
PLOT$colour$term$bowel_prep <- "#542788"
PLOT$colour$term$abx_pre_fmt <- "#b2abd2"
PLOT$colour$term$fmt_sample_prep.fresh <- "#e31a1c"
PLOT$colour$term$fmt_route.nasal <- "#fb9a99"
PLOT$colour$term$richness.donor.scaled <- "#6a3d9a"
PLOT$colour$term$richness.pre.scaled <- "#cab2d6"
PLOT$colour$term$richness.ratio.donor_pre <- "#a6dba0"
PLOT$colour$term$dissimilarity.donor_pre <- "#ff7f00"
PLOT$colour$term$dissimilarity.post_pre <- "#fdbf6f"
PLOT$colour$term$abd_ratio.donor_pre <- "#33a02c"
PLOT$colour$term$allele_dist.donor_pre <- "#b2df8a"

#Colours by study
PLOT$colour$study <- list()
PLOT$colour$study$FAME <- "#1f78b4"
PLOT$colour$study$FATLOSE <- "#1f78b4"
PLOT$colour$study$TURN <- "#6a3d9a"
PLOT$colour$study$FMT_AU <- "#e31a1c"
PLOT$colour$study$Damman <- "#ff7f00"
PLOT$colour$study$Nusbaum <- "#fdbf6f"
PLOT$colour$study$Lee <-  "#fb9a99"
PLOT$colour$study$Vaughn <- "#1f78b4"
PLOT$colour$study$Smillie <- "#a6cee3"
PLOT$colour$study$Kumar <- "#33a02c"
PLOT$colour$study$Moss <- "#fb9a99"
```

Define an order of FMTs based on study & indication.

```{r}
data.fmt %>%
  arrange(study.full, fmt.type, indication, clinical.response.summary, fmt.id) %>%
  pull(fmt.id) -> order.fmt
```

Define a set of species to consider in tests, based on a minimum number of scored FMTs (20 by default, but include other cutoffs as well).

```{r}
#Get a list of species
dat.transmission.scored %>%
  #Drop problematic FMTs
  #filter(! fmt.id %in% drop.fmt.id) %>%
  filter(! (fmt.id %in% c("FMT_Vaughn.02", "FMT_Vaughn.07") & timepoint.fmt == 56)) %>%
  filter(! (fmt.id %in% c("FAME.06b") & timepoint.fmt == 28)) %>%
  #Select only properly scored FMTs
  filter(! outcome %in% c("not observed", "autologous FMT")) %>%
  #Remove redundant Collinsella species
  filter(! species %in% drop.Collinsella) %>% #group_by(fmt.id, timepoint.fmt) %>% add_tally() %>% group_by(fmt.id, timepoint.fmt, outcome) %>% dplyr::summarise(ratio = n() / n) %>% distinct() %>% arrange(desc(ratio))
  #Select only FMT-species combinations where the species was present in the donor (otherwise, no engraftment possible...)
  filter(inc.donor) %>%
  #Select only last FMT time point
  group_by(fmt.id) %>%
  top_n(1, timepoint.fmt) %>%
  #Select only species scored in >=20 FMTs
  group_by(species) %>%
  dplyr::summarise(n = n()) -> tmp

#Apply filters, also kick out problematic "Vampirovibrionia" sp.
tmp %>% filter(n >= 20 & !species %in% c("meta_mOTU_v2_5475", "meta_mOTU_v2_7007")) %>% pull(species) -> use.species.engraftment
tmp %>% filter(n >= 30 & !species %in% c("meta_mOTU_v2_5475", "meta_mOTU_v2_7007")) %>% pull(species) -> use.species.engraftment.gt_30
tmp %>% filter(n >= 50 & !species %in% c("meta_mOTU_v2_5475", "meta_mOTU_v2_7007")) %>% pull(species) -> use.species.engraftment.gt_50
```

Pre-organise taxonomic annotations for relevant species.

```{r}
data.phenotype.gmm %>%
  filter(mOTU %in% use.species.engraftment) %>%
  dplyr::mutate(
    phylum = fct_relevel(phylum, c(
      "Firmicutes_A", "Firmicutes", "Firmicutes_C",
      "Bacteroidota", "Actinobacteriota",
      "Proteobacteria", "Desulfobacterota",
      "Verrucomicrobiota", "Methanobacteriota"
    )),
    class = fct_relevel(class, c(
      "Clostridia", "Clostridia_A",
      "Bacilli",
      "Negativicutes",
      "Bacteroidia",
      "Coriobacteriia",
      "Actinomycetia",
      "Gammaproteobacteria", "Alphaproteobacteria",
      "Desulfovibrionia",
      "Verrucomicrobiae", "Lentisphaeria",
      "Methanobacteria"
    )),
    order = fct_relevel(order, c(
      "Clostridiales", "Oscillospirales", "Lachnospirales", "TANB77", "Peptostreptococcales", "Monoglobales", "Monoglobales_A",
      "Christensenellales",
      "Lactobacillales", "Erysipelotrichales", "RF39", "Haloplasmatales", "Acholeplasmatales", "ML615J-28",
      "Veillonellales", "Acidaminococcales", "Selenomonadales",
      "Bacteroidales",
      "Coriobacteriales",
      "Actinomycetales",
      "Burkholderiales", "Enterobacterales",
      "RF32",
      "Desulfovibrionales",
      "Verrucomicrobiales", "Victivallales",
      "Methanobacteriales"
    ))
  ) -> tmp

#Family (by decreasing count w/in groups)
tmp %>% group_by(phylum, class, order, family) %>% tally() %>% arrange(phylum, class, order, desc(n)) %>% pull(family) -> tmp.order.fam
tmp %>% mutate(family = fct_relevel(family, tmp.order.fam)) -> tmp

#Genus
tmp %>% group_by(phylum, class, order, family, genus) %>% tally() %>% arrange(phylum, class, order, family, desc(n)) %>% pull(genus) -> tmp.order.gen
tmp %>% dplyr::mutate(genus = fct_relevel(genus, tmp.order.gen[tmp.order.gen != ""])) -> tmp

data.phenotype.gmm.gt_20 <- tmp
```

Store parameters in common file.

```{r}
save(
  PARAM,
  PLOT,
  drop.Collinsella,
  order.fmt,
  use.species.engraftment,
  use.species.engraftment.gt_30,
  use.species.engraftment.gt_50,
  data.phenotype.gmm.gt_20,
  file = paste0(PARAM$folder.data, "param.analysis_transmission.Rdata")
)
```

## Summarise GMGC-based GMM profiles.

```{r}
dat.gmm.gmgc %>%
  left_join(data.gmm, by = c("gmm" = "Module")) %>%
  dplyr::mutate(
    module_manual = case_when(
      gmm == "MF0103" ~ "mucin_degradation",
      gmm == "MF0086" ~ "acetate_production",
      gmm %in% c("MF0093", "MF0094", "MF0095") ~ "propionate_production",
      gmm %in% c("MF0088", "MF0089") ~ "butyrate_production",
      TRUE ~ gmm
    )
  ) %>%
  group_by(sample, module_manual) %>%
  summarise(count = sum(count), count_norm = sum(count_norm)) %>%
  group_by(sample) %>%
  mutate(count.clr = log2(count / mean(count))) %>%
  ungroup() %>%
  dplyr::filter(module_manual %in% c("mucin_degradation", "acetate_production", "propionate_production", "butyrate_production")) -> dat.gmm.gmgc.tmp

dat.gmm.gmgc.summarised %>%
  dplyr::filter(HL1 %in% c("amino acid degradation", "carbohydrate degradation", "lipid degradation") & Position == "input") %>%
  rename(module_manual = HL1) %>%
  dplyr::mutate(
    module_manual = case_when(
      module_manual == "carbohydrate degradation" ~ "saccharolysis",
      module_manual == "amino acid degradation" ~ "proteolysis",
      module_manual == "lipid degradation" ~ "lipolysis"
    )
  ) %>%
  bind_rows(dat.gmm.gmgc.tmp) -> tmp

tmp %>%
  dplyr::select(-Position, -count_norm, -count) %>%
  pivot_wider(names_from = module_manual, values_from = count.clr) -> dat.gmm.gmgc.summarised.wide

tmp %>%
  dplyr::select(-Position, -count.clr, -count) %>%
  pivot_wider(names_from = module_manual, values_from = count_norm) -> dat.gmm.gmgc.summarised.wide.norm
```       

Store.

```{r}
save(dat.gmm.gmgc.summarised.wide, dat.gmm.gmgc.summarised.wide.norm, file = PARAM$file.gmm_gmgc_wide)
```

## Calculate phylogenetic complementarity for each species in each sample

Re-prep tree (subset to relevant species).

```{r}
#Order species by phylogeny, stratified by general engraftment success
tmp.label <- data.phenotype.gmm %>% filter(mOTU %in% use.species.engraftment) %>% pull(label)
tmp.tree <- phytools::bind.tip(tree.full, "Bacteroides ovatus (A0023)", where = which(tree.full$tip.label == "Bacteroides ovatus (r1073)"), edge.length = 0.01)
tmp.tree <- phytools::bind.tip(tmp.tree, "Streptococcus parasanguinis_B (A0144)", where = which(tmp.tree$tip.label == "Streptococcus parasanguinis (r0144)"), edge.length = 0.01)
tmp.tree <- phytools::bind.tip(tmp.tree, "Blautia_A sp900066355 (A2800)", where = which(tmp.tree$tip.label == "Blautia_A sp900066355 (r4719)"), edge.length = 0.01)
tmp.tree <- phytools::bind.tip(tmp.tree, "Bifidobacterium adolescentis (A0399)", where = which(tmp.tree$tip.label == "Bifidobacterium adolescentis (r1156)"), edge.length = 0.01)
tmp.tree <- phytools::bind.tip(tmp.tree, "Enterocloster clostridioformis (A3455)", where = which(tmp.tree$tip.label == "Enterocloster clostridioformis (r0979)"), edge.length = 0.01)
tree.engraft.gt_20 <- ape::keep.tip(tmp.tree, tmp.label)
tree.engraft.gt_20$tip.label <- use.species.engraftment[match(tree.engraft.gt_20$tip.label, tmp.label)]
```

Calculate all species' pairwise cophenetic distances.

```{r}
mat.cophenetic <- ape::cophenetic.phylo(tree.engraft.gt_20)
mat.cophenetic <- mat.cophenetic[use.species.engraftment, use.species.engraftment]
```

Calculate the `phylogenetic complementarity` of each species in each sample:

* minimum phylogenetic (cophenetic) distance to the closest relative in that sample
* forced to 0 if species itself is present

```{r}
load(paste0(PARAM$folder.data, "data.coverage.COG.MAG.Rdata"))

#Get incidence of relevant novel species
dat.cov.COG %>%
  filter(species %in% use.species.engraftment) %>%
  dplyr::mutate(inc = ifelse(ver_cov.median > 0, 1, 0)) %>%
  select(-n_cog, -ver_cov.mean, -ver_cov.sd, -ver_cov.median) -> tmp

bind_rows(tmp, tibble(species = "ANI_AL_0023", sample_alias = data.sample %>% filter(! sample_alias %in% tmp$sample_alias & sample_alias %in% colnames(data.mOTU)) %>% pull(sample_alias), inc = 0)) %>%
  filter(sample_alias %in% data.sample$sample_alias) %>%
  pivot_wider(names_from = "sample_alias", values_from = "inc", values_fill = 0) %>%
  column_to_rownames("species") %>%
  as.matrix() -> data.mOTU.inc.ANI95

#Tranform mOTU table to incidence & combine info
data.inc <- rbind(
  1 * (data.mOTU[rownames(data.mOTU) %in% use.species.engraftment, ] > 0),
  data.mOTU.inc.ANI95[, colnames(data.mOTU)]
)
data.inc <- data.inc[use.species.engraftment, ]
data.inc[data.inc == 0] <- 1000

mat.phylo_complementarity <- matrix(0, nrow = nrow(data.inc), ncol = ncol(data.inc), dimnames = dimnames(data.inc))
for (s in use.species.engraftment) {
  tmp <- (mat.cophenetic[s, ] + 0.01) * data.inc
  tmp[s, data.inc[s, ] > 1] <- 1000
  mat.phylo_complementarity[s, ] <- apply(tmp, 2, min) - 0.01
}

as_tibble(mat.phylo_complementarity, rownames = "species") %>%
  pivot_longer(cols = -species, names_to = "sample", values_to = "phylo.complementarity") -> dat.phylo_complementarity

save(dat.phylo_complementarity, file = paste0(PARAM$folder.data, "data.phylo_complementarity.Rdata"))
```


## Prepare transmission data for processing.

Collate and reorganise all relevant variables:

* technical variables (seq depth)
* clinical variables (indication, FMT procedure, etc.)
* community-level microbiome variables (richness, community similarity, etc.)
* species-level microbiome variables (presence/absence pre FMT, abundance ratio, allele distance donor-pre)
* species phenotypes (oxygen tolerance, etc.)

Precalculate and scale minimum seq depth per FMT.

```{r}
seq_depth.mean <- mean(data.sample$base_count, na.rm=T)
seq_depth.sd <- sd(data.sample$base_count, na.rm=T)
```

Reformat `dist.gmm.gmgc` to `tibble`.

```{r}
dist.gmm.gmgc %>%
  as.matrix() %>%
  as_tibble(rownames = "sample_1") %>%
  pivot_longer(cols = -sample_1, names_to = "sample_2", values_to = "dist") -> tmp

dist.gmm.gmgc.tbl <-  bind_rows(
  tmp,
  tmp %>% mutate(sample_tmp = sample_1, sample_1 = sample_2, sample_2 = sample_tmp) %>% select(-sample_tmp)
)

```

Consolidate data across FMTs.

```{r}
dat.transmission.scored %>%
  #Drop problematic FMTs
  filter(! (fmt.id %in% c("FMT_Vaughn.02", "FMT_Vaughn.07") & timepoint.fmt == 56)) %>%
  filter(! (fmt.id %in% c("FAME.06b") & timepoint.fmt == 28)) %>%
  #Drop autologous FMTs and non-scored FMTs
  filter(! outcome %in% c("not observed", "autologous FMT")) %>%
  filter(inc.donor | inc.pre | inc.post) %>%
  #Subset to species of interest
  filter(species %in% use.species.engraftment) %>%
  filter(!species %in% drop.Collinsella) %>%
  ###########################
  #Re-quantify outcome data
  ###########################
  #=> continuous/fractional outcomes
  dplyr::mutate(
    engraftment = case_when(
      n.pos.scored < 100 ~ 0,
      !inc.donor ~ 0,
      outcome == "rejection novel" ~ 0,
      TRUE ~ donor.specific
    ),
    frac.don = case_when(
      !inc.donor ~ 0,
      outcome %in% c("species lost", "species retained", "rejection novel", "influx novel") ~ 0,
      TRUE ~ donor.specific
    ),
    frac.rec = case_when(
      !inc.pre ~ 0,
      outcome %in% c("species lost", "engraftment novel", "novel unclear", "influx novel") ~ 0,
      TRUE ~ recipient.specific
    ),
    frac.coexist = case_when(
      !inc.pre ~ 0,
      !inc.donor ~ 0,
      !inc.post ~ 0,
      outcome %in% c("species lost", "species retained", "rejection novel", "engraftment novel", "influx novel", "novel unclear", "autologous FMT") ~ 0,
      TRUE ~ donor.specific + recipient.specific
    ),
    sp_retained = case_when(
      outcome == "species retained" ~ recipient.specific,
      outcome == "autologous FMT" & inc.pre ~ recipient.specific,
      TRUE ~ 0
    ),
    reject_consp = ifelse(outcome == "rejection conspecific", recipient.specific, 0),
    coexist_consp.rec = ifelse(outcome %in% c("coexistence", "engraftment conspecific", "influx conspecific"), recipient.specific, 0),
    coexist_consp.donor = ifelse(outcome %in% c("coexistence", "rejection conspecific", "influx conspecific"), donor.specific, 0),
    engraft_consp = ifelse(outcome == "engraftment conspecific", donor.specific, 0),
    reject_novel = ifelse(outcome == "rejection novel", 1, 0),
    engraft_novel = ifelse(outcome %in% c("engraftment novel", "novel unclear"), donor.specific, 0),
    sp_lost = case_when(
      outcome == "species lost" ~ 1,
      outcome == "autologous FMT" & inc.pre & !inc.post ~ 1,
      TRUE ~ 0
    ),
    influx_consp = case_when(
      outcome == "species retained" ~ 1 - recipient.specific,
      outcome == "autologous FMT" & inc.pre & inc.post ~ 1 - recipient.specific,
      outcome == "autologous FMT" & !inc.pre & inc.post ~ 1,
      outcome == "autologous FMT" & inc.pre & !inc.post ~ 0,
      outcome %in% c("rejection conspecific", "engraftment conspecific", "coexistence", "influx conspecific") ~ 1 - recipient.specific - donor.specific,
      outcome == "species lost" ~ 0,
      TRUE ~ 0
    ),
    influx_novel = case_when(
      outcome == "influx novel" ~ 1,
      outcome %in% c("novel unclear", "engraftment novel") ~ 1 - donor.specific,
      TRUE ~ 0
    ),
    influx_total = influx_consp + influx_novel,
  ) %>%
  #=> binarised outcomes
  dplyr::mutate(
    oc.donor_takeover = ifelse(inc.donor & inc.post & outcome %in% c("engraftment novel", "engraftment conspecific"), 1, 0),
    oc.donor_engraftment = ifelse(inc.donor & inc.post & outcome %in% c("engraftment novel", "engraftment conspecific", "coexistence"), 1, 0),
    oc.donor_rejection = ifelse(inc.donor & outcome %in% c("rejection novel", "rejection conspecific", "influx novel", "influx conspecific"), 1, 0),
    oc.rec_persistence = ifelse(inc.pre & inc.post & outcome %in% c("species retained", "rejection conspecific"), 1, 0),
    oc.rec_resilience = ifelse(inc.pre & outcome %in% c("species retained", "rejection conspecific", "coexistence"), 1, 0),
    oc.rec_species_loss = ifelse(inc.pre & outcome %in% c("species lost"), 1, 0),
    oc.rec_turnover = ifelse(inc.post & outcome %in% c("engraftment novel", "engraftment conspecific", "influx novel", "influx conspecific", "novel unclear"), 1, 0)
  ) %>%
  ###########################
  #Prepare the various variables to be tested later on
  ###########################
  #Compute per-species variables
  mutate(
    #Calculate ratio of relative abundances of species in donor and pre-FMT sample
    #=> use a pseudo-count to avoid Inf
    abd_ratio.donor_pre = case_when(
      is.na(abd.pre) | is.na(abd.donor) ~ as.numeric(NA),
      TRUE ~ log2((abd.donor + 10^-5) / (abd.pre + 10^-5))
    ),
    #Conspecific vs novel
    #=> was species present pre-FMT?
    present_pre_FMT = ifelse(inc.pre, 1, 0)
  ) %>%
  #Allele distance between donor and pre-FMT sample
  left_join(
    bind_rows(dat.allele_distance.fmt, dat.allele_distance.fmt %>% rename(sample_2 = sample_1, sample_1 = sample_2)) %>% dplyr::select(species, fmt.id, sample_1, sample_2, dist),
    by = c("species" = "species", "fmt.id" = "fmt.id", "sample.donor" = "sample_1", "sample.pre" = "sample_2")
  ) %>%
  dplyr::rename(allele_dist.donor_pre = dist) %>%
  group_by(species) %>%
  mutate(
    allele_dist.donor_pre = case_when(
      #Force distance to 1 if species wasn't present pre-FMT
      !inc.donor | !inc.pre ~ 1,
      #Interpolate mean allele distance if NA
      is.na(allele_dist.donor_pre) ~ mean(allele_dist.donor_pre, na.rm=T),
      #Otherwise use computed distance
      TRUE ~ allele_dist.donor_pre
    )
  ) %>%
  ungroup() %>%
  #Allele diversity
  #=> donor samples
  left_join(dat.allele_diversity.filtered %>% dplyr::select(species, sample, allele_div, allele_div.norm), by = c("species" = "species", "sample.donor" = "sample")) %>%
  rename(allele_div.donor = allele_div, allele_div.donor.norm = allele_div.norm) %>%
  #=> pre-FMT samples
  left_join(dat.allele_diversity.filtered %>% dplyr::select(species, sample, allele_div, allele_div.norm), by = c("species" = "species", "sample.pre" = "sample")) %>%
  rename(allele_div.pre = allele_div, allele_div.pre.norm = allele_div.norm) %>%
  #=> post-FMT samples
  left_join(dat.allele_diversity.filtered %>% dplyr::select(species, sample, allele_div, allele_div.norm), by = c("species" = "species", "sample.post" = "sample")) %>%
  rename(allele_div.post = allele_div, allele_div.post.norm = allele_div.norm) %>%
  #Phylogenetic complementarity
  #=> minimum phylogenetic distance of focal species to recipient pre-FMT community
  left_join(dat.phylo_complementarity, by = c("species" = "species", "sample.pre" = "sample")) %>%
  #Metabolic complementarity
  #=> average novelty scores of species KO compositiion vs resident metabolic setup
  left_join(dat.ko_complementarity %>% dplyr::select(species, fmt.id, novelty.avg) %>% rename(ko.complementarity = novelty.avg), by = c("species", "fmt.id")) %>%
  ###########################
  #Microbiome community-level factors
  #=> donor richness
  left_join(data.sample %>% dplyr::select(sample_alias, q.0) %>% dplyr::mutate(q.scaled = (q.0 - mean(q.0, na.rm=T)) / sd(q.0, na.rm=T)), by = c("sample.donor" = "sample_alias")) %>%
  dplyr::rename(richness.donor = q.0, richness.donor.scaled = q.scaled) %>%
  #=> richness pre-FMT
  left_join(data.sample %>% dplyr::select(sample_alias, q.0) %>% dplyr::mutate(q.scaled = (q.0 - mean(q.0, na.rm=T)) / sd(q.0, na.rm=T)), by = c("sample.pre" = "sample_alias")) %>%
  dplyr::rename(richness.pre = q.0, richness.pre.scaled = q.scaled) %>%
  #=> richness post-FMT
  left_join(data.sample %>% dplyr::select(sample_alias, q.0) %>% dplyr::mutate(q.scaled = (q.0 - mean(q.0, na.rm=T)) / sd(q.0, na.rm=T)), by = c("sample.post" = "sample_alias")) %>%
  dplyr::rename(richness.post = q.0, richness.post.scaled = q.scaled) %>%
  #=> richness ratio
  mutate(richness.ratio.donor_pre = log2(richness.donor / richness.pre)) %>%
  mutate(richness.ratio.post_pre = log2(richness.post / richness.pre)) %>%
  #=> donor functional redundancy
  left_join(data.sample %>% dplyr::select(sample_alias, fr.ent) %>% dplyr::mutate(fr.scaled = (fr.ent - mean(fr.ent, na.rm=T)) / sd(fr.ent, na.rm=T)), by = c("sample.donor" = "sample_alias")) %>%
  dplyr::rename(fr.donor = fr.ent, fr.donor.scaled = fr.scaled) %>%
  #=> pre-FMT functional redundancy
  left_join(data.sample %>% dplyr::select(sample_alias, fr.ent) %>% dplyr::mutate(fr.scaled = (fr.ent - mean(fr.ent, na.rm=T)) / sd(fr.ent, na.rm=T)), by = c("sample.pre" = "sample_alias")) %>%
  dplyr::rename(fr.pre = fr.ent, fr.pre.scaled = fr.scaled) %>%
  #=> post-FMT functional redundancy
  left_join(data.sample %>% dplyr::select(sample_alias, fr.ent) %>% dplyr::mutate(fr.scaled = (fr.ent - mean(fr.ent, na.rm=T)) / sd(fr.ent, na.rm=T)), by = c("sample.post" = "sample_alias")) %>%
  dplyr::rename(fr.post = fr.ent, fr.post.scaled = fr.scaled) %>%
  #=> Bray-Curtis similarity donor-pre
  mutate(dissimilarity.donor_pre = beta.div[["Bray_Curtis"]][cbind(sample.donor, sample.pre)]) %>%
  #=> Bray-Curtis similarity post-pre (strength of overall shift)
  mutate(dissimilarity.post_pre = beta.div[["Bray_Curtis"]][cbind(sample.post, sample.pre)]) %>%
  #GMM profiles
  #=> summarised gmm profiles
  left_join(dat.gmm.gmgc.summarised.wide %>% rename_with(~paste0(. , ".donor"))) %>%
  left_join(dat.gmm.gmgc.summarised.wide %>% rename_with(~paste0(. , ".pre"))) %>%
  left_join(dat.gmm.gmgc.summarised.wide %>% rename_with(~paste0(. , ".post"))) %>%
  #=> GMM-based "metabolic" distances between donor and recipient
  left_join(dist.gmm.gmgc.tbl, by = c("sample.donor" = "sample_1", "sample.pre" = "sample_2")) %>% rename(dist_gmm.donor_pre = dist) %>%
  #=> GMM-based "metabolic" distances between donor and recipient
  left_join(dist.gmm.gmgc.tbl, by = c("sample.post" = "sample_1", "sample.pre" = "sample_2")) %>% rename(dist_gmm.post_pre = dist) %>%
  ###########################
  #Technical variables
  #=> minimum base depth among donor, pre and post FMT samples
  left_join(data.sample %>% dplyr::select(sample_alias, base_count), by = c("sample.donor" = "sample_alias")) %>% dplyr::rename(depth.donor = base_count) %>%
  left_join(data.sample %>% dplyr::select(sample_alias, base_count), by = c("sample.pre" = "sample_alias")) %>% dplyr::rename(depth.pre = base_count) %>%
  left_join(data.sample %>% dplyr::select(sample_alias, base_count), by = c("sample.post" = "sample_alias")) %>% dplyr::rename(depth.post = base_count) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(min_depth = ifelse(is.na(depth.donor) & is.na(depth.pre) & is.na(depth.post), NA, min(depth.donor, depth.pre, depth.post, na.rm=T))) %>%
  ungroup() %>%
  dplyr::mutate(min_depth = (min_depth - seq_depth.mean) / seq_depth.sd) %>%
  ###########################
  #Taxa phenotype data
  left_join(data.taxonomy.gtdb %>% dplyr::select(mOTU, label), by = c("species" = "mOTU")) %>%
  left_join(data.phenotype.gmm %>% dplyr::select(mOTU, size.genome, motility, sporulation, oxygen_pref, saccharolysis, proteolysis, lipolysis, mucin_degradation, acetate_production, propionate_production, butyrate_production, butyrate.type), by = c("species" = "mOTU")) %>%
  mutate(
    motility = case_when(motility == "motile" ~ 1, motility == "non-motile" ~ 0, TRUE ~ as.numeric(NA)),
    sporulation = case_when(sporulation == "sporulating" ~ 1, sporulation == "non-sporulating" ~ 0, TRUE ~ as.numeric(NA)),
    oxygen.facultative = case_when(oxygen_pref %in% c("aerobic", "facultative") ~ 1, oxygen_pref == "anaerobic" ~ 0, TRUE ~ as.numeric(NA)),
    butyrate_production = case_when(butyrate_production == 0 & butyrate.type %in% c("neither", "inferred_only") ~ 0, TRUE ~ 1),
    acetate_production = ifelse(acetate_production > 0, 1, 0),
    propionate_production = ifelse(propionate_production > 0, 1, 0)
  ) %>%
  ###########################
  #Clinical data
  left_join(data.fmt %>% dplyr::select(fmt.id, study.full, indication, fmt.route, fmt.sample_prep, bowel.prep.summary, vancomycin, fidaxomycin, antibiotics.summary, clinical.response.summary), by = "fmt.id") %>%
  mutate(
    indication.esbl = ifelse(indication == "antibiotics_resistance", 1, 0),
    indication.mets = ifelse(indication == "metabolic_syndrome", 1, 0),
    indication.ibs = ifelse(indication == "irritable_bowel_syndrome", 1, 0),
    indication.uc = ifelse(indication == "ulcerative_colitis", 1, 0),
    indication.cd = ifelse(indication == "morbus_crohn", 1, 0),
    indication.ibd = ifelse(indication %in% c("morbus_crohn", "ulcerative_colitis"), 1, 0),
    indication.rcdi = ifelse(indication == "recurrent_c_diff", 1, 0),
    indication.mel = ifelse(indication == "melanoma", 1, 0),
    indication.ren = ifelse(indication == "renal_carcinoma", 1, 0),
    indication.tou = ifelse(indication == "tourette_syndrome", 1, 0),
    indication.ctr = ifelse(indication == "healthy_volunteer", 1, 0),
    fmt_route.nasal = ifelse(fmt.route %in% c("nasoduodenal tube", "nasogastric tube", "colonoscopy and nasoduodenal tube"), 1, 0),
    fmt_route.rectal = ifelse(fmt.route %in% c("colonoscopy", "colonoscopy and enema", "retention enema for 5d", "colonoscopy and nasoduodenal tube", "olonoscopy and oral capsule"), 1, 0),
    fmt_sample_prep.fresh = ifelse(fmt.sample_prep == "fresh", 1, 0),
    bowel_prep = ifelse(bowel.prep.summary, 1, 0),
    abx_pre_fmt = ifelse(antibiotics.summary, 1, 0),
    clinical_success = ifelse(clinical.response.summary == "responder", 1, 0)
  ) %>%
  distinct() -> dat.transmission.scored.outcome

dat.transmission.scored.outcome %>% filter(inc.donor) -> dat.transmission.scored.engraftment
```

```{r}
save(dat.transmission.scored.outcome, dat.transmission.scored.engraftment, file = PARAM$file.data_transmission.scored.consolidated)
```






